tests/filter.json

```diff
@@ -478,6 +490,30 @@
         [42]
       ]
     },
+    {
+      "name": "on primitive, selects nothing",
+      "selector": "$.a[?@ == 1]",
+      "document": {"a": 1},
+      "result": []
+    },
+    {
+      "name": "on primitive object member value, selects nothing",
+      "selector": "$[?@.a == 1]",
```

Comment:
This and the following three tests seem to be checking that a relative path (`@<...>`) can have the same path syntax as a global path (`$<...>`).  If that's something we're testing, that's fine, but I'd argue that we'd need to almost replicate the entire suite for relative paths.

Reply:
This one was based on https://github.com/hiltontj/serde_json_path/issues/49 where the reporter assumed `@` refers to the JSON object instead of the member value.

And the tests "name selector on array" and "index selector on object" are supposed to cover the same issue in `serde_json_path` where the underlying bug was that an unexpected type (which should select nothing) erroneously always passed the filter. I assume such a bug pattern could affect other libraries as well.

Reply:
> where the reporter assumed @ refers to the JSON object instead of the member value.

Two things here:

1. We shouldn't be interested in testing what an end user expects, even (especially) if that expectation is wrong.  We should be testing _implementation_ behavior per the spec.  If the implementation is doing it right and the user expected something different, then the solution is to correct the user.  We don't need a test for it.
2. If we _are_ going to test this, then we should test for interpreting `@` directly; we don't need the `.a`.  `"$[?@ == 1]" would be a much more targeted test.

Reply:
> 1. We shouldn't be interested in testing what an end user expects, even (especially) if that expectation is wrong.  We should be testing _implementation_ behavior per the spec.

Fair point. But I think here if the document is `{"a": 1}`, then it is necessary to have `.a`. If you used `$[?@ == 1]`, then `[1]` is the correct result.

But if you add `.a` (i.e. `?@.a == 1`), then you should get no result. However, libraries might get this wrong if they silently ignore the `.a` because it cannot be applied to a non-object. And with https://github.com/hiltontj/serde_json_path/issues/49 there is a precedent for that exact issue.

Reply:
I view this differently.  The _test_ should be that `@` represents the `1`.  This is the simple test that verifies the implementation has the correct behavior.

Separately, we test that `.a` doesn't apply to a non-object (which is your primitive tests).

That `.a` can't be applied to a non-object in the context of `@` is merely a _consequence_ of the correct behavior and doesn't specifically need a test.

Reply:
`singular-query`s, i.e., relative paths with `@`, are defined separately in the standard ABNF. So, this leads implementations to have separate parsing, AST representations, evaluation mechanisms, etc. - as was the case for `serde_json_path`.

Therefore, I think there is value in having tests that ensure the behaviour of queries against primitives for _both_ absolute _and_ relative paths. Relative paths can only appear in filters, so having separate tests for those requires queries like @Marcono1234 has used here.

---

I suppose that the motivation for this test case is that one implementor (me) made a logic error in the handling of singular queries, so, having this test would prevent another from doing the same. It is not possible to come up with tests for every conceivable logic error, or misinterpretation of JSONPath, but I don't see the harm in having a test that guards against one that _did happen_ (and the `.a` was relevant in that case).

Reply:
> `singular-querys`, i.e., relative paths with @, are defined separately in the standard ABNF.

While true, `@`-paths can also be existence tests, which can handle all selector constructs.  I believe (on mobile now; can verify letter) that we have some treats that ensure only singular paths are used in comparisons, though.  If not, we should add them in another PR.

> I suppose that the motivation for this test case is that one implementor (me) made a logic error in the handling of singular queries, so, having this test would prevent another from doing the same.

Okay. That this was an implementation error wasn't understood.  I thought this was a user that expected something incorrect.

However, I still maintain, that this test is overly complex and can be broken down into smaller constituent pieces that still achieve the goal of verifying proper operation.

Reply:
Sorry for taking so long to get back...

> While true, @-paths can also be existence tests

This escaped me, thanks for pointing that out. I was incorrect there. Relative path `!=` singular query. The singular query defined in the ABNF is only relevant to comparisons. Relative paths in existence tests or as function arguments can be non-singular.